runner/local/queueing_runner.go

```diff
@@ -0,0 +1,228 @@
+)
+
+// This runner manages multiple (possibly concurrent) run and status requests.
+// One request is run at a time, additional requests are queued till the current request is finished running.
+
+// QueueingRunner is initialized with the runner to use to run the requests, the maximum number of requests to
+// manage on the queue, and the 'callback' channel the runner will use to tell queueing runner that it has
+// completed the last request.
+//
+// When a request is received it is immediately added to the queue.  As the runner becomes available, if
+// if there is are requests on the queue, the top request on the queue is sent to the runner
+//
+// If the max queue length is exceed, QueueingRunner rejects any new request till length of the queue is below
+// the max.
+//
+// QueueingRunner assigns a unique id (run id) to each new request.  The Run() function returns this id and a
+// status indicating if the request has been successfully queued or if the request has been denied (due to queue
+// length exceeded or some other error).
+
+// The status of a request is obtained with the run id.
+
+const QueueFullMsg = "No resources available. Please try later."
+const UnspportedFeatureMsg = "Not implemented.  This feature is not available for QueueingRunner"
+
+// The request queue entries are commandAndId structs
+type commandAndId struct {
+	id  runner.RunId
+	cmd *runner.Command
+}
+
+// requestResponse structs are returned in the Run() and Status() response channels
+type requestResponse struct {
+	status *runner.ProcessStatus
+	err    error
+}
+
+//TODO - do we really need a new channel for each run request?
+// This struct to associates a response channel to each Run() request
+type runRequestEvent struct {
+	cmd               *runner.Command
+	onQueueResponseCh chan requestResponse // chan to use to indicate that the request is queued
+}
+
+type QueueingRunner struct {
+	runner            runner.Runner
+	done              bool
+	maxQueueLen       int
+	nextRunId         int // used to assign unique runIds to the run requests
+	runQueue          []commandAndId
+	runnerIsAvailable bool
+
+	notifyRunnerAvailCh chan struct{} // the runner will use this channel to signal it is available
+
+	runRequestsCh chan runRequestEvent // channel for synchronizing calls to Run()
+
+	statusRequestsCh chan string          // channel for synchronizing status requests
+	statusResponseCh chan requestResponse // channel for unblocking the status requests
+
+	// map the queueing runner run ids to the runner's run ids. We need runner's run id for Status()
+	queueIdToRunnerIdMap map[string]string
+}
+
+// This must be used to initialize QueueingRunner properly
+func NewQueuingRunner(context context.Context,
+	runner runner.Runner,
+	maxQueueLen int,
+	runnerAvailableCh chan struct{}) runner.Runner {
+
+	qRunner := &QueueingRunner{
+		runner:               runner,
+		done:                 false,
+		maxQueueLen:          maxQueueLen,
+		nextRunId:            0,
```

Comment:
we should talk about this... IMHO QueueingRunner should become PoolRunner.  we should not have queueing in runners that really run the commands.